zilliqa/src/consensus.rs

```diff
@@ -684,13 +702,55 @@ impl Consensus {
             .transpose()
     }
 
-    pub fn get_transaction_receipt(&self, hash: Hash) -> Result<Option<TransactionReceipt>> {
+    pub fn get_block_hash_from_transaction(&self, tx_hash: &Hash) -> Result<Option<Hash>> {
+        self.block_hash_reverse_index
+            .get(tx_hash.0)?
+            .map(|ivec| ivec.as_ref().try_into())
+            .transpose()
+    }
+
+    pub fn get_transaction_receipts_in_block(
+        &self,
+        block_hash: Hash,
+    ) -> Result<Option<Vec<TransactionReceipt>>> {
         self.transaction_receipts
-            .get(hash.0)?
-            .map(|encoded| Ok(bincode::deserialize::<TransactionReceipt>(&encoded)?))
+            .get(block_hash.0)?
+            .map(|encoded| Ok(bincode::deserialize::<Vec<TransactionReceipt>>(&encoded)?))
             .transpose()
     }
 
+    pub fn get_transaction_receipt(&self, hash: &Hash) -> Result<Option<TransactionReceipt>> {
+        let Some(block_hash) = self.get_block_hash_from_transaction(hash)? else {
+            return Ok(None);
+        };
+        let Some(block_receipts) = self.get_transaction_receipts_in_block(block_hash)? else {
+            return Ok(None);
+        };
+        Ok(block_receipts
+            .into_iter()
+            .find(|receipt| receipt.tx_hash == *hash))
+    }
+
+    pub fn get_logs_in_block(&self, hash: Hash, event: Event) -> Result<Vec<Log>> {
+        let Some(receipts) = self.get_transaction_receipts_in_block(hash)? else {
+            return Ok(vec![]); // no transations in block, most likely
+        };
+
```

Comment:
I just reminded myself of how this works and currently we only calculate the block's bloom filter in the API layer when we're returning a block, so we don't have it available at this point to filter.

We might want to change this in the future though to actually persisting the bloom filters in blocks, because it probably has quite a few use-cases in terms of efficiency (e.g. in #433)

Reply:
Good idea but we're not even storing the bloom filter on the blocks yet - I'll file it away under #436 and continue working on shards for now.